### PR TITLE
Issue #15025: Add support for LITERAL_WHEN in WhiteSpaceAfterCheck

### LINKED ISSUE #15025: Add Check Support for Java 21 Pattern Matching for Switch Syntax: WhitespaceAfter
child of #14961 

I have read check documentation: https://checkstyle.org/checks/whitespace/whitespaceafter.html#WhitespaceAfter
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```

➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="WhitespaceAfter"/>
    </module>
</module>
➜  src cat Test.java 
record ColoredPoint(int p, int x, String c) { }

public class Test {

    void test(Object obj) {
        int y = 0;
        if(y>=2) { }            // violation
        switch (obj) {
            case ColoredPoint(int x, _, _) when(x >= 2) -> {}     // expected violation

            default -> {}
        }
    }
}
➜  src javac --enable-preview --release 21 Test.java                        
Note: Test.java uses preview features of Java SE 21.
Note: Recompile with -Xlint:preview for details.
➜  src java -jar checkstyle-10.18.1-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:7:9: 'if' is not followed by whitespace. [WhitespaceAfter]
Audit done.
Checkstyle ends with 1 errors.


```
---

**Describe what you want in detail**

violation on `when` because there is no white space after when and before the parens

### PR DESCRIPTION
**Pull request by mahfouz72**
closes #15025 

Diff Regression config: https://gist.githubusercontent.com/mahfouz72/afd57aa6e51ca161b7927acd959e5211/raw/759257e4c69433ba141d623872b21d1c3dd3a4f3/check.xml
Diff Regression patch config: https://gist.githubusercontent.com/mahfouz72/33967bf86e4a29cafa429771b0e95808/raw/49ec3a90de0ff4ebd87979177c2cda622f751c4e/check_patch.xml

### CODE REVIEW

**Comment by mahfouz72** [comment]
github, generate report

**Comment by mahfouz72** [comment]
github, generate site

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java
Comment (nrmancuso) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java:17]: Please add a case where we negate this expression (`when!(...`)

**Review by nrmancuso — CHANGES_REQUESTED** [review]

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java:17]: done

**Review by nrmancuso — APPROVED** [review]
I am good, but please investigate why we are missing checks (should be ~111).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheck.java
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheck.java:144]: @nrmancuso I see we made this token default in this check, but not in the other check.

File: src/main/resources/google_checks.xml
Comment (rnveach) [inline src/main/resources/google_checks.xml:113]: @mahfouz72 @nrmancuso @romani @Zopsss Same as other PR.
We need to show this follows google's style.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java
Comment (rnveach) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java:3]: Same as other PR. Show examples with new lines, and show enable/disable of property `allowLineBreaks`.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/main/resources/google_checks.xml
Comment (Zopsss) [reply src/main/resources/google_checks.xml:113]: my answer would be same as https://github.com/checkstyle/checkstyle/pull/15311#discussion_r1686824725. The style guide does not have rules for java 21 keywords but we're talking about basic spacing rules here, so I think we should add support for this token, it'll improve overall code readability.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java:4]: done. 

>  enable/disable of property allowLineBreaks.

what property?

**Comment by mahfouz72** [comment]
Failure is unrelated 

`[ERROR] Failed to execute goal com.groupon.maven.plugin.json:json-schema-validator:1.2.0:validate (default) on project checkstyle: Exception while locating input file; file=https://raw.githubusercontent.com/oasis-tcs/sarif-spec/main/sarif-2.1/schema/sarif-schema-2.1.0.json: Could not find resource 'https://raw.githubusercontent.com/oasis-tcs/sarif-spec/main/sarif-2.1/schema/sarif-schema-2.1.0.json'. -> [Help 1]`

the [resource](https://raw.githubusercontent.com/oasis-tcs/sarif-spec/main/sarif-2.1/schema/sarif-schema-2.1.0.json) is on the web actually

File: src/main/resources/google_checks.xml
Comment (romani) [reply src/main/resources/google_checks.xml:113]: Please run Google formatter over Input to be double confident that space is required.
I think space should be there

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java
Comment (Zopsss) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java:4]: We have [allowLineBreaks](https://checkstyle.sourceforge.io/checks/whitespace/nowhitespacebefore.html#Properties) in NoWhitespaceBefore check but we don't have it for WhitespaceAfter check.

File: src/main/resources/google_checks.xml
Comment (Zopsss) [reply src/main/resources/google_checks.xml:113]: I have did it here: https://github.com/checkstyle/checkstyle/pull/15311#discussion_r1686824725

Formatter does adds space around `when`

Comment (romani) [reply src/main/resources/google_checks.xml:113]: @mahfouz72 , please add few examples (copy from regular test Inputs is ok) of "when" to new Input at folder
https://github.com/checkstyle/checkstyle/tree/master/src/it/resources-noncompilable/com/google/checkstyle/test/chapter4formatting
and create test method at https://github.com/checkstyle/checkstyle/blob/master/src/it/java/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/HorizontalWhitespaceTest.java

this Check is in 4.6.2 https://checkstyle.org/google_style.html#a4.6.2

same request was done at https://github.com/checkstyle/checkstyle/pull/15311 , it might required to be added ones.

Comment (mahfouz72) [reply src/main/resources/google_checks.xml:113]: https://github.com/checkstyle/checkstyle/pull/15359

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java
Comment (rnveach) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespaceafter/InputWhitespaceAfterLiteralWhen.java:4]: ok, I may have mixed up PRs or documentation.

File: src/main/resources/google_checks.xml
Comment (rnveach) [reply src/main/resources/google_checks.xml:113]: I am awaiting responses from @romani @nrmancuso on this. If this agreed to covered by the style guide, then we need to add examples to the IT inputs.

Comment (mahfouz72) [reply src/main/resources/google_checks.xml:113]: @rnveach please see 2 comments above yours. I think @romani agrees and I already asked to extend IT inputs  there is a PR

Comment (romani) [reply src/main/resources/google_checks.xml:113]: I agree to extend google style config.

**Comment by mahfouz72** [comment]
@romani @rnveach @nrmancuso please decide what to do to unblock this

**Comment by romani** [comment]
I placed my approval.
I am to extend Google style config.

**Comment by mahfouz72** [comment]
@rnveach @nrmancuso Then please let's get this across merge. extend it input already has a PR but blocked by this PR

File: src/main/resources/google_checks.xml
Comment (nrmancuso) [reply src/main/resources/google_checks.xml:113]: Ok, we will extend the google style config, I have removed the blocked label.

Comment (rnveach) [reply src/main/resources/google_checks.xml:113]: ok. I don't understand why https://github.com/checkstyle/checkstyle/pull/15359 is a separate PR. It should be merged in with this since this is where it is being added to google config.

Comment (mahfouz72) [reply src/main/resources/google_checks.xml:113]: > since this is where it is being added to google config.

but it is being added in https://github.com/checkstyle/checkstyle/pull/15311 also and we use both checks (whitespaceAfter and around) in ITs input. That's why I make a separate PR to extend input after merging both checks updates

Comment (nrmancuso) [reply src/main/resources/google_checks.xml:113]: I am good with the separate PR, let's get this across the finish line

**Review by rnveach — APPROVED** [review]

**Pull request merged**
